blockstore/apps/bundles/models.py

```diff
@@ -54,13 +54,22 @@
 """
 import uuid
 
+from django.conf import settings
 from django.db import models
 
 from .store import DraftRepo, SnapshotRepo, bytes_from_hex_str
 
 MAX_CHAR_FIELD_LENGTH = 180
 
 
+"""
+Sqlite is used for testing on edx-platform, but doesn't contain the utf8mb4_general_ci collation sequence.
+
+So we detect this case here, and use a different db_collation if we're running on Sqlite.
+"""
+DB_COLLATION = 'binary' if 'sqlite' in settings.DATABASES['default']['ENGINE'] else 'utf8mb4_general_ci'
```

Comment:
@symbolist @gabor-boros  There's a database compatibility problem with the `db_collation` setting that was chosen for some of the fields in these models.

Blockstore tests and production use MySQL, so using `db_collation='utf8mb4_general_ci'` was fine.
However, my attempts to run the unit tests with blockstore installed have failed because edx-platform uses sqlite.

* Sqlite, doesn't support `utf8mb4_general_ci`; the best it can do is [`binary`](https://www.sqlite.org/datatype3.html#collating_sequences).
* Postgres doesn't support `utf8mb4_general_ci` either, and its use contradicts the [Postgres compatibility as documented](https://github.com/openedx/blockstore/pull/149/files#diff-2d364908a68c64f2a417d1036d2c3e5f95d8c4e1f585757818ee31b09ce79564R20-R240) for these models. Postgres's closest built-in equivalent collation sequence is the [standard `ucs_basic`](https://www.postgresql.org/docs/10/collation.html#id-1.6.10.4.5.5).

While it's relatively easy to configure the charset and collation sequence options at the `settings.DATABASES` level, we don't want to do this for all fields, and there [isn't a nice way to do this per-field](https://stackoverflow.com/questions/59503476/setting-collation-and-charset-per-model-table-in-django-and-mysql), except by modifying the model code directly.

So a full solution would require:
* detecting [which database is in use for the given class](https://stackoverflow.com/a/30459399/4302112), and 
* returning an appropriate collation setting for that database engine.

All of which makes the model.py and the generated migrations pretty complicated, and difficult to reliably test and maintain.

@ormsbee has raised the very timely issue of whether we can formally [drop support for Postgres](https://discuss.openedx.org/t/postgresql-and-open-edx/6441), but that's a longer discussion.

Any suggestions for how to handle this better right now?

Reply:
Does it make sense to create a custom field that subclasses `CharField` and has the logic to apply database-specific collations during migrations? I think it's primarily two use cases:

* Human-readable text that should be collated to be case-insensitive and as aware of accents, etc. as possible.
* Identifiers, which should be case sensitive.

Reply:
I _think_ the tests don't depend on unicode-aware cast insensitive sorting, so `binary` should work fine for test purposes? The main thing they test is that all unicode characters including emojis can be used without issue, and `binary` should be fine for that.

And yeah +1 to what Dave is suggesting.

Reply:
@bradenmacdonald 

> I think the tests don't depend on unicode-aware cast insensitive sorting, 

Yep, this hack works for testing, but it's not pretty, and doesn't handle the Postgres use case.

> Does it make sense to create a custom field that subclasses CharField and has the logic to apply database-specific collations during migrations?

Interesting suggestion, @ormsbee.. a custom field class seems doable, but it will still need to be passed information about the database selected for the migration so that it can exhibit the best field options for that engine. I'd have to dig into Django's field and migration logic to see if that can be done with the existing auto-generated migrations or not, but it's worth a try!

Since this is likely something that would be broadly useful outside of Blockstore, is there an `openedx` repo that would be an appropriate place for this change to live?

I'm going to have to leave this hack in place right now though, and create a follow-up task for this, since it's out of scope for this change.

Reply:
> Since this is likely something that would be broadly useful outside of Blockstore, is there an openedx repo that would be an appropriate place for this change to live?

[edx-django-utils](https://github.com/openedx/edx-django-utils) is probably what you want. It's installed on pretty much every service.